object-store/src/main/java/org/hypertrace/config/objectstore/IdentifiedObjectStoreWithFilter.java

```diff
@@ -0,0 +1,65 @@
+package org.hypertrace.config.objectstore;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.hypertrace.config.service.change.event.api.ConfigChangeEventGenerator;
+import org.hypertrace.config.service.v1.ConfigServiceGrpc;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+
+/**
+ * IdentifiedObjectStoreWithFilter is an abstraction over {@link IdentifiedObjectStore} to allow
+ * filtering on fetched configs
+ *
+ * @param <T> Config Object
+ * @param <F> Filter
+ */
+public abstract class IdentifiedObjectStoreWithFilter<T, F> extends IdentifiedObjectStore<T> {
+  protected IdentifiedObjectStoreWithFilter(
+      ConfigServiceGrpc.ConfigServiceBlockingStub configServiceBlockingStub,
+      String resourceNamespace,
+      String resourceName,
+      ConfigChangeEventGenerator configChangeEventGenerator) {
+    super(configServiceBlockingStub, resourceNamespace, resourceName, configChangeEventGenerator);
+  }
+
+  public List<T> getAllFilteredConfigs(RequestContext context, F filter) {
+    Stream<T> configObjectsStream = getAllObjects(context).stream().map(ConfigObject::getData);
+    if (!isFilterEmpty(filter)) {
+      configObjectsStream =
+          configObjectsStream
+              .map(configObject -> filterConfig(configObject, filter))
+              .filter(Optional::isPresent)
+              .map(Optional::get);
+    }
+    return configObjectsStream.collect(Collectors.toUnmodifiableList());
+  }
+
+  public Optional<T> getFilteredConfig(RequestContext context, String contextId, F filter) {
+    Optional<T> fetchedConfig = getData(context, contextId);
+    if (!isFilterEmpty(filter)) {
+      return fetchedConfig.flatMap(config -> filterConfig(config, filter));
+    }
+    return fetchedConfig;
+  }
+
+  /**
+   * Method to obtain filtered config A config object can consist of multiple configs, e.g. array,
+   * which can be partially filtered and returned. In cases where filter applies to entire config
```

Comment:
Can you point me to an example of a config that applies a filter internally like this?

Reply:
`ai.traceable.anomaly.config.service.detector.anomalydetection.AnomalyDetectionConfigManagerImpl`

Reply:
ahhh - this one is very different than pretty much every other config. So it gets read/written as one massive object? Was there a particular reason it was done this way, or is it just tech debt?

Will read on, but if this is the only config that buries multiple independent configs in one object, we can/should consider giving it a custom store so as not to affect the API for the straightforward cases.